### DISCUSSION TITLE
How to do dynamic themes in Android

### DISCUSSION THREAD

**Discussion by Ghostbird** [dotnet/maui#14062] [Q&A]
I was trying to [dynamically apply a theme](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/theming?view=net-maui-7.0#consume-theme-resources) to my Android app, and ran into some issues:

Using e.g. `Shell.BackgroundColor={DynamicResource Primary}` doesn't work. The title bar becomes a colour which is defined nowhere, neither in MAUI, nor in the Android system's theme settings.

Buttons etc. don't change colour because their colour is set in `Platforms/Android/Resources/values/colors.xml`. If I remove `colors.xml`, some buttons in Android will not use (static) MAUI themes but instead use the Android system theme.

How do I correctly theme the app? It seems the documentation posits a theoretical _happy flow_ functionality, which doesn't work in practice.

Here's [my example set-up](https://github.com/Ghostbird/DynamicResource)

**Comment by AndreasReitberger** [comment]
In order to update the `PrimaryColor`, I'm using a `Service`.
https://github.com/AndreasReitberger/SharedMauiCoreLibrary/blob/main/src/SharedMauiCoreLibrary/Services/PlatformThemeService.cs

```cs
#if ANDROID || IOS || MACCATALYST || WINDOWS
    public partial class PlatformThemeService
    {
        // Based on: https://stackoverflow.com/a/39164921/10083577
        public partial void SetStatusBarColor(Color color);
    }
#endif
```

On Android, it looks like that.

```cs
public partial class PlatformThemeService
    {
        // Source: https://stackoverflow.com/a/39164921/10083577
        public partial void SetStatusBarColor(Color color)
        {
            // The SetStatusBarcolor is new since API 21
            if (Build.VERSION.SdkInt >= BuildVersionCodes.Lollipop)
            {
                var androidColor = color.AddLuminosity(-0.1f).ToAndroid();
                Microsoft.Maui.ApplicationModel.Platform.CurrentActivity?.Window.SetStatusBarColor(androidColor);
            }
            else
            {
                // Here you will just have to set your 
                // color in styles.xml file as shown below.
            }
        }
    }
```

Usage in my MAUI App. I call this method whenever I change the `PrimaryColor`

```cs
/// <summary>
        /// Updates the `PrimaryColor` for platform specific resources, as the StatusBar
        /// </summary>
        /// <param name="theme">The ThemeColorInfo to be changed to</param>
        /// <param name="app">The current app</param>
        public void UpdatePlatformThemeColor(ThemeColorInfo theme) {
            try {
                new PlatformThemeService()?.SetStatusBarColor(theme?.PrimaryColor);
            }
            catch (Exception exc) {
                EventManager.Instance.LogError(exc);
            }
        }
```

In the `MauiProgram.cs` add this event, so that the colors also get updated on startup.

```cs
            .ConfigureLifecycleEvents(events =>
            {
#if ANDROID
                events.AddAndroid(android => android
                    // The statusbar color needs to be updated after the activity has been created
                    .OnCreate((activity, bundle) => UpdateThemeColor(activity, bundle))
                    );
                static void UpdateThemeColor(Activity activity, Bundle bundle)
                {
                    new PlatformThemeService().SetStatusBarColor(App.Instance.ThemeColor?.PrimaryColor);
                }
#endif
            })
```

Maybe this helps you :)

**Reply by Ghostbird** [reply]
This sounds helpful, but shouldn't his be done automatically by MAUI? And I have no indication that this will solve the button colours too.
